### PR TITLE
[Feature] Support automatic partition by value

### PR DESCRIPTION
**Pull request by Astralidea**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 104 / 166 (62.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|39|68|57.35%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 174 (63.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|46|76|60.53%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862, 882]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -424,48 +438,60 @@ Status OlapTablePartitionParam::find_tablets(Chunk* chunk, std::vector<OlapTable
...
-                                    << row.debug_string();
-                            (*partition_not_exist_row_values)[0].emplace_back(column->debug_item(i));
-                        }
+                if (is_list_partition) {
+                    // list partition
+                    auto it = _partitions_map.find(&row);
+                    if (it != _partitions_map.end() && _part_contains(it->second, &row)) {
+                        (*partitions)[i] = it->second;
+                        (*indexes)[i] = (*indexes)[i] % it->second->num_buckets;
+                        LOG(INFO) << "find list partition:" << it->second->id << " row:" << row.debug_string();
```
Comment (ABingHuang) [inline be/src/exec/tablet_info.cpp:447]: remove it

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 174 (63.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|46|76|60.53%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862, 882]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 174 (63.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|46|76|60.53%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862, 882]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 174 (63.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|46|76|60.53%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862, 882]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java [hunk 2]
```diff
@@ -764,25 +773,125 @@ public static String parseLiteralExprToDateString(LiteralExpr expr, PrimitiveTyp
...
+                        partitionName, Collections.singletonList(partitionValue), partitionProperties);
+
+                multiItemListPartitionDesc.setSystem(true);
+                AddPartitionClause addPartitionClause =
+                        new AddPartitionClause(multiItemListPartitionDesc, distributionDesc,
+                                partitionProperties, false);
+                result.put(partitionName, addPartitionClause);
+            }
+        } else {
+            throw new AnalysisException("automatic partition only support expression range partition.");
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java:862]: can support range or list partition now?

**Review by packy92 — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 174 (63.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|3|33.33%|[426, 427]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|2|50.00%|[290]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|25|48|52.08%|[1744, 1822, 1823, 1824, 1849, 1850, 1852, 1853, 1854, 1855, 1856, 1857, 1858, 1859, 1862, 1863, 1864, 1865, 1866, 1867, 1868, 1871, 1872]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|46|76|60.53%|[785, 789, 796, 800, 807, 811, 815, 822, 837, 838, 839, 840, 842, 843, 844, 845, 846, 847, 848, 849, 850, 852, 853, 855, 856, 859, 860, 861, 862, 882]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|19|25|76.00%|[425, 426, 427, 428, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/ast/ListPartitionDesc.java|4|5|80.00%|[260]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/service/PartitionMeasure.java|6|6|100.00%|[]|

**Comment by Astralidea** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**